source/extensions/formatter/metadata/metadata.cc

```diff
@@ -0,0 +1,80 @@
+#include "source/extensions/formatter/metadata/metadata.h"
+
+#include <string>
+
+#include "source/common/config/metadata.h"
+#include "source/common/formatter/substitution_formatter.h"
+#include "source/common/http/utility.h"
+#include "source/common/protobuf/utility.h"
+#include "source/common/runtime/runtime_features.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Formatter {
+
+// Metadata formatter for route's metadata.
+class RouteMetadataFormatter : public ::Envoy::Formatter::MetadataFormatter {
+public:
+  RouteMetadataFormatter(const std::string& filter_namespace, const std::vector<std::string>& path,
+                         absl::optional<size_t> max_length)
+      : ::Envoy::Formatter::MetadataFormatter(filter_namespace, path, max_length,
+                                              [](const StreamInfo::StreamInfo& stream_info)
+                                                  -> const envoy::config::core::v3::Metadata* {
+                                                auto route = stream_info.routeEntry();
+
+                                                if (route == nullptr) {
+                                                  return nullptr;
+                                                }
+                                                return &route->metadata();
+                                              }) {}
```

Comment:
somewhat weird formatting, but it's probably fine and it's not worth creating the lambda from a private method.